spinand.c

```diff
@@ -0,0 +1,279 @@
+ *   3. Read Cache
+ */
+
+static int spi_nand_read_page(struct flashctx *flash, unsigned int row_addr)
+{
+	uint8_t cmd[1 + SPINAND_ROW_ADDR_LEN] = { SPINAND_READ_PAGE, };
+
+	int addr_len = spi_nand_prepare_row_address(cmd, row_addr);
+
+	return spi_send_command(flash, 1 + addr_len, 0, cmd, NULL);
+}
+
+static int spi_nand_wait_for_page(struct flashctx *flash, unsigned int row_addr)
+{
+	int ret;
+
+	ret = spi_nand_read_page(flash, row_addr);
+	if (ret)
+		return ret;
+	return spi_nand_wait(flash);
+}
+
+static int spi_nand_read_cache(struct flashctx *flash, unsigned int column_addr,
+                               uint8_t *bytes, unsigned int len)
+{
+	uint8_t cmd[1 + SPINAND_COLUMN_ADDR_LEN] = { SPINAND_READ_CACHE, };
+
+	int addr_len = spi_nand_prepare_column_address(cmd, column_addr);
+
+	/* Send Read */
+	return spi_send_command(flash, 1 + addr_len + DUMMY_BYTE, len, cmd, bytes);
+}
+
+static int spi_nand_inpage_read(struct flashctx *flash, unsigned int row_addr, unsigned int column_addr,
+                                uint8_t *bytes, unsigned int len, unsigned int chunksize)
+{
+	/* Read data within one page */
+	int ret;
+
+	if (chunksize == 0)
+		chunksize = len;
+	ret = spi_nand_wait_for_page(flash, row_addr);
+	if (ret)
+		return ret;
+
+	while (len) {
+		unsigned int data_to_read = min(chunksize, len);
+		ret = spi_nand_read_cache(flash, column_addr, bytes, data_to_read);
+		if (ret)
+			return ret;
+		len -= data_to_read;
+		column_addr += data_to_read;
+		bytes += data_to_read;
+		printf("page %d\n",len);
+	}
+
+	return 0;
+}
+
+static inline unsigned int ilog2(unsigned int x) {
+	int res = 0;
+	while (x >>= 1)
+		++res;
+	return res;
+}
+
+int spi_nand_read_chunked(struct flashctx *flash, uint8_t *buf, unsigned int start,
+                          unsigned int len, unsigned int chunksize)
+{
+	/* Read data over multiple pages */
+	int ret;
+
+	/* Calculate acture column address len */
+	unsigned int column_addr_len = ilog2(flash->chip->page_size);
+	unsigned int column_addr_mask = ~(-1 << column_addr_len);
+	printf("%d %d %d\n",flash->chip->page_size,column_addr_len,column_addr_mask);
+
+	while (len) {
+		unsigned int data_to_read = min(chunksize, len);
+		ret = spi_nand_inpage_read(flash, start >> column_addr_len, start & column_addr_mask, buf, len, data_to_read);
+		if (ret)
+			return ret;
+		len -= data_to_read;
+		start += data_to_read;
+		buf += data_to_read;
+		printf("chunked %d\n",len);
+	}
+
+	return 0;
+}
+
+/***************************************
+ * PROGRAM CYCLE
+ ***************************************/
+
+
+/***************************************
+ * PROBE
+ ***************************************/
+
+static char *strncpy_tospace(char *dest, const char *src, unsigned int len)
+{
+	int i;
+
+	for (i = 0; i < len; i++) {
+		dest[i] = src[i];
+		if (dest[i] == ' ') {
+			dest[i] = '\0';
+			break;
+		}
+	}
+	return dest;
+}
+
+static void probe_spi_nand_toshiba(struct flashctx *flash, const uint8_t *parameters)
```

Comment:
Suggest referencing doc(s) if implementing manufacturer specific implementations.

If their READ ID comes back with ONFI then I think its fair to honor ONFI spec.

An issue I've run into is that vendors will specify ONFI revision 0. This is not a valid specification. The real spec starts at 1.0. I take this to mean that they haven't fully implemented the spec. I think warning users of this fact and requiring a `force` option would be sufficient in these cases.